File: examples/keyboard_input_interrupt.py
from main import *


def main_run() -> None:
    # 1. Create a keyboard device that triggers interrupt vector 1 (RST 1 -> vector 0x0008)
    kbd = KeyboardDevice(interrupt_vector=1)

    # 2. Attach keyboard device to machine on Port 0x01
    machine = Machine.create(
        address_lines=16,
        data_lines=8,
        devices=[(kbd, [0x01])]
    )
    cpu = machine.cpu
    ram = machine.ram
    cpu.reg_sp.write(0x1000)

    # 3. Write ISR for RST 1 at 0x0008:
    # 0x0008: IN 0x01   - Read ASCII key value from keyboard port into A
    # 0x000A: MOV B, A  - Save input key in Register B
    # 0x000B: RET       - Return from interrupt handler
    ram.write(Mem(0x0008), Data.byte(Opcode.IN))
    ram.write(Mem(0x0009), Data.byte(0x01))
    ram.write(Mem(0x000A), Data.byte(Opcode.MOV_B_A))
    ram.write(Mem(0x000B), Data.byte(Opcode.RET))

    # 4. Main Program: Enable interrupts and NOP in loop
    program = Program([
        Instruction(Opcode.EI),  # Enable Interrupts (inte = True)
        Instruction(Opcode.NOP),
        Instruction(Opcode.HLT)
    ])

    machine.load(program, Mem(0x00A0))

    # Trigger a keyboard keypress event
    kbd.trigger_key_press('K')

    # Emulate the hardware interrupt request asserting INTR pin
    cpu.intr = True

    # 5. Run the machine
    machine.run()

    print("Keyboard Input & Interrupt Example State:")
    print(f"Register B: {hex(cpu.reg_b.value)} (Expected: 0x4b - ASCII 'K')")
    print(f"Interrupts Enabled (inte): {cpu.inte}")

if __name__ == "__main__":
    main_run()

File: examples/keyboard_input_interrupt.rs
//! Ported from the reference project's `examples/keyboard_input_interrupt.py`.
//! Run with `cargo run --example keyboard_input_interrupt`.

use emu8085::{Addr, Instruction, KeyboardDevice, Machine, Opcode, Program};

fn main() {
    // 1. Create a keyboard device that triggers interrupt vector 1 (RST 1 -> vector 0x0008)
    let mut kbd = KeyboardDevice::with_vector(1);

    // Trigger a keyboard keypress event
    kbd.press_char('K').unwrap();

    // 2. Create machine and attach keyboard device to port 0x01
    let mut machine = Machine::create(16, 8);
    machine.attach_device(Box::new(kbd), &[0x01]);
    machine.cpu.regs.sp = Addr(0x1000);

    // 3. Write ISR for RST 1 at 0x0008:
    // 0x0008: IN 0x01   - Read ASCII key value from keyboard port into A
    // 0x000A: MOV B, A  - Save input key in Register B
    // 0x000B: RET       - Return from interrupt handler
    machine.ram.write(Addr(0x0008), Opcode::IN as u8);
    machine.ram.write(Addr(0x0009), 0x01);
    machine.ram.write(Addr(0x000A), Opcode::MOV_BA as u8);
    machine.ram.write(Addr(0x000B), Opcode::RET as u8);

    // 4. Main Program: Enable interrupts and NOP in loop
    let program = Program::new(vec![
        Instruction::new(Opcode::EI),  // Enable Interrupts (inte = true)
        Instruction::new(Opcode::NOP),
        Instruction::new(Opcode::HLT),
    ]);

    machine.load(&program, Addr(0x00A0)).expect("program compiles");

    // Emulate the hardware interrupt request asserting INTR pin
    machine.cpu.intr = true;

    // 5. Run the machine
    machine.run();

    let cpu = &machine.cpu;
    println!("Keyboard Input & Interrupt Example State:");
    println!("Register B: 0x{:02X} (Expected: 0x4B - ASCII 'K')", cpu.regs.b);
    println!("Interrupts Enabled (inte): {}", cpu.inte);

    assert_eq!(cpu.regs.b, b'K');
}
